### PR TITLE
Fix narrow on empty tensors after negative size support.

### PR DESCRIPTION
**Pull request by gchanan**

### CODE REVIEW

**Review by zou3519 — APPROVED** [review]
lgtm, thanks for catching this @gchanan!

I'm not sure about the error message but that can be fixed later

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -154,7 +154,9 @@ Tensor &as_strided_(Tensor& self, IntList size, IntList stride) {
 Tensor narrow(const Tensor& self, int64_t dim, int64_t start, int64_t length) {
   AT_CHECK(self.dim() > 0, "narrow() cannot be applied to a 0-dim tensor.");
   auto cur_size = self.size(dim);
-  start = maybe_wrap_dim(start, cur_size);
+  if (start != cur_size) {  // start being the end is valid, but not a valid dim specification.
+    start = maybe_wrap_dim(start, cur_size);
```
[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/native/TensorShape.cpp]: Yah, I don't think using `maybe_wrap_dim` is a good idea here; it's just not the right abstraction.  But this was the minimal change to get things working again.

**Comment by bhushan23** [comment]
can we also add test case for start = cur_size in test_narrow?

**Comment by gchanan** [comment]
We have one if you compile with zero size dim support `CFLAGS="-DUSE_TH_SIZE_ZERO_DIM"` -- it's a little tricky without that because it would only be valid for a 1-dimensional tensor.

**Review by ezyang — APPROVED** [review]

**Comment by bhushan23** [comment]
@gchanan MemongerTest.test_simple_memonger is asserting

**Comment by gchanan** [comment]
that's a caffe2 test, I don't think it's related.

**Pull request closed without merging**